programs/vote/src/vote_state/mod.rs

```diff
@@ -870,9 +891,33 @@ impl VoteState {
         // `new_vote_state` passed all the checks, finalize the change by rewriting
         // our state.
         if self.root_slot != new_root {
-            // TODO to think about: Note, people may be incentivized to set more
-            // roots to get more credits, but I think they can already do this...
-            self.increment_credits(epoch);
+            // Award vote credits based on the number of slots that were voted on and have reached finality
+            if feature_set
+                .map(|feature_set| feature_set.is_active(&feature_set::timely_vote_credits::id()))
+                .unwrap_or(false)
+            {
+                if let Some(in_slot) = in_slot {
+                    // Credits are awarded based on the "timeliness" of the vote transaction which landed them.  For
+                    // each finalized slot, there was one voted-on slot in the new vote state that was responsible for
```

Comment:
Not always true consider something like this:
Current on chain vote state:
`Root 0, Votes: 1, 2, 3, ... 31`

Some time later the local tower looks like this but vote txs haven't been landing:
`Root 7, Votes: 8, ... 31, 32, 33, 34, 35, 36, 37, 38`

We hit threshold check and end up popping off some votes (vote txs still not landing properly)
`Root 7, Votes: 8, ... 31, 32, 33, 34, 52`

Finally we submit this new vote state and the tx lands. We've "finalized" 7 slots from our old vote state (slots 1 through 7) but not all of the votes that were responsible for it are still in our new vote state.

Reply:
But in this situation the "current state" will be:

Root: 0, Votes: 1...31

And the "new state" being supplied by VoteStateUpadte will be:

Root: 7, Votes: 8, ... 31, 32, 33, 34, 52

So the logic will detect that there are 7 votes that were the oldest votes in the current state that are not in the new state, and will then count 7 slots has having finalized.  It will then compute credits based on the timeliness of the "newest" 7 slots in the new state.

It is true that the votes that actually caused the deque are not represented in the new state.  So the credits will be awarded based on the current newest vote slots at the time of the new vote state update.  This creates a proxy for vote timeliness - more timely votes will always receive more credits, but "how many" more credits is in fact a function of how much expiry has gone on in the interim.

Reply:
What will happen is that the local tower's concept of credits awarded for the local vote state will be different from what is actually recorded.  But that's OK as the local tower's credits number is not used for anything, its value is just a side-effect of tower bookkeeping.